### PR TITLE
Add the bond, angle, and dihedral table potentials to the v3 API.

### LINKED ISSUE #360: Port Table potentials to v3 API
HOOMD's tabulated potential implementation and API are being very widely used. Their implementation, and Python API in particular, could use a thorough review and be improved if possible.

### PR DESCRIPTION
**Pull request by joaander**
## Description

Add the bond, angle, and dihedral table potentials to the v3 API.

## Motivation and context

Researchers that utilize these features should be able to upgrade to v3.

Resolves #360

## How has this been tested?

I added Table tests to the relevant unit tests file. In the process, I refactored the unit tests so it is easier for developers to manage test cases.

## Change log

```
Added:

* ``md.bond.Table`` - Tabulated bond potential.
* ``md.angle.Table`` - Tabulated angle potential.
* ``md.dihedral.Table`` - Tabulated dihedral potential.

Changed:

* [breaking] - renamed ``md.angle.Cosinesq`` to ``md.angle.CosineSquared``.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

File: hoomd/md/TableAngleForceCompute.cc [hunk 1]
```diff
@@ -9,9 +9,6 @@
     \brief Defines the TableAngleForceCompute class
 */
 
-#undef VT0
-#undef VT1
-
```
Comment (tommy-waltmann) [inline hoomd/md/TableAngleForceCompute.cc:14]: Why were these removed?

File: hoomd/md/TableDihedralForceCompute.cc [hunk 2]
```diff
@@ -12,9 +12,6 @@
 
 using namespace std;
 
-#undef VT0
-#undef VT1
-
```
Comment (tommy-waltmann) [inline hoomd/md/TableDihedralForceCompute.cc:17]: Same as above

**Review by tommy-waltmann — COMMENTED** [review]
Overall looks good, just a small question in addition to the ones below:

What happened to the tabulated angle potential that was in v2? For the tabulated bond and dihedral potentials, an old class was removed to make a new class for v3, but I don't see such a class removed to make way for the v3 tabulated angle potential.

File: hoomd/md/TableAngleForceCompute.cc (hunk 1 above)
Comment (joaander) [reply hoomd/md/TableAngleForceCompute.cc:14]: They are unnecessary. Nowhere does the code `#define VT0`  or `#define VT1`.

**Comment by joaander** [comment]
> Overall looks good, just a small question in addition to the ones below:
> 
> What happened to the tabulated angle potential that was in v2? For the tabulated bond and dihedral potentials, an old class was removed to make a new class for v3, but I don't see such a class removed to make way for the v3 tabulated angle potential.

The `angle.table` class was not there to be removed. I did not check the git history to see when it was removed or who removed it.

**Review by tommy-waltmann — APPROVED** [review]

File: hoomd/md/BondTablePotential.h [hunk 3]
```diff
@@ -68,6 +68,18 @@ class PYBIND11_EXPORT BondTablePotential : public ForceCompute
...
+    void setParamsPython(std::string type, pybind11::dict params);
+
+    /// Get the parameters for a particular type.
+    pybind11::dict getParams(std::string type);
+
+    /// Get the width
+    unsigned int getWidth()
+        {
+        return m_table_width;
+        }
```
Comment (ispivack) [inline hoomd/md/BondTablePotential.h:81]: I was under the impression that it is bad form to include implementation in the header in order to not accidentally violate the "one definition rule". (I am new to C++, so this could be incorrect)

Comment (joaander) [reply hoomd/md/BondTablePotential.h:81]: C++ developers commonly include the implementation for short methods in the header file. This allows the compiler to inline those functions when they are called for performance. When the implementation is in the .cc file, there is no opportunity for inlining and the method must be called with all the overhead of a function call.

We add preprocessor macros near the top of the file to create an inclusion guard. This ensures that the method is only defined once.
